subsys/net/lib/sockets/sockets_select.c

```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sys/time.h>
+
+#include <net/socket.h>
+
+/* Outputs results in predefined word_idx and bit_mask vars. */
+#define FD_SET_CALC_OFFSETS(set) { \
+	word_idx = fd / (sizeof(set->bitset[0]) * 8); \
+	unsigned int b_idx = fd % (sizeof(set->bitset[0]) * 8); \
+	bit_mask = 1 << b_idx; \
+	}
+
+void ZSOCK_FD_ZERO(zsock_fd_set *set)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(set->bitset); i++) {
+		set->bitset[i] = 0;
+	}
+}
+
+int ZSOCK_FD_ISSET(int fd, zsock_fd_set *set)
+{
+	uint32_t word_idx, bit_mask;
+
+	if (fd < 0 || fd >= ZSOCK_FD_SETSIZE) {
+		return 0;
+	}
+
+	FD_SET_CALC_OFFSETS(set);
+
+	return (set->bitset[word_idx] & bit_mask) != 0;
+}
+
+void ZSOCK_FD_CLR(int fd, zsock_fd_set *set)
+{
+	uint32_t word_idx, bit_mask;
+
+	if (fd < 0 || fd >= ZSOCK_FD_SETSIZE) {
+		return;
+	}
+
+	FD_SET_CALC_OFFSETS(set);
+
+	set->bitset[word_idx] &= ~bit_mask;
+}
+
+void ZSOCK_FD_SET(int fd, zsock_fd_set *set)
+{
+	uint32_t word_idx, bit_mask;
+
+	if (fd < 0 || fd >= ZSOCK_FD_SETSIZE) {
+		return;
+	}
+
+	FD_SET_CALC_OFFSETS(set);
+
+	set->bitset[word_idx] |= bit_mask;
+}
+
+int zsock_select(int nfds, zsock_fd_set *readfds, zsock_fd_set *writefds,
+		 zsock_fd_set *exceptfds, struct timeval *timeout)
+{
+	struct pollfd pfds[CONFIG_NET_SOCKETS_POLL_MAX];
+	int i, res, poll_timeout;
+	int num_pfds = 0;
+	int num_selects = 0;
+	int fd_no;
+
+	fd_no = 0;
+	for (i = 0; i < ARRAY_SIZE(((zsock_fd_set *)0)->bitset); i++) {
```

Comment:
> It will work I assume? Took me a moment to catch up what's going on here.

Sure, should work, unless weird typo went in somewhere ;-). That's how `offsetof()` is implemented usually. And all routed in the lack of an operator to go inside a type (vs inside a value of a type), like `::` in C++.

It's used in 2 places actually. Dunno if a macro would help, will consider that.

Reply:
(Can't use any of readfds/writefds/execepfds directly, as any of them can be NULL).

Reply:
> And all rooted in the lack of an operator to go inside a type (vs inside a value of a type), like ::

So, I added a macro, but it covers just one use, with ARRAY_SIZE, because another uses just normal sizeof().

Of course, all that stuff can be rewritten to avoid all that at all, by introducing an extra global constant to define the size of array. The whole idea was to use a local nameless expression to define this size, and use "true" (array)sizeofs elsewhere. Maybe it's not worth it.

Reply:
> I'm fine with how it's done now.

Good to hear.

Anyway, writing to catch myself on a goof:

> (Can't use any of readfds/writefds/execepfds directly, as any of them can be NULL).

So, I complain that any of them can be null, and then use explicit null instead %) ;-).